### PR TITLE
Faster radix sort

### PR DESCRIPTION
**Pull request by LSchwerdt**
Unrolling the loop in Base.Sort.radix_sort_pass! increases performance by about 50%.
(on my Ryzen 3900x)

This makes the code somewhat less clean, but in my opinion the performance benefit is worth it.

Disclaimer: I did not fully build Julia including this PR (yet). The change is very small and I did test the modified function, but I do not want to setup everything to build Julia if the PR might be rejected based on a choice of code clarity vs. performance.

[Here is a simple test script](https://github.com/LSchwerdt/MiscJulia/blob/4c2c9c3abe04bcb5dd9cdcadb841ed64442958fd/radix_pass_unroll/radixpassunroll.jl) to try out the changed function in isolation.
This is the output:
```
2000/20000
...
20000/20000

base:     5.063 seconds
unrolled: 3.052 seconds
speedup factor: 1.66
Test Passed
```

### CODE REVIEW

File: base/sort.jl [hunk 1]
```diff
@@ -1112,6 +1112,13 @@ function radix_sort!(v::AbstractVector{U}, lo::Integer, hi::Integer, bits::Unsig
         shift < bits || return true
     end
 end
+
+macro repeat4x(a)
+    quote
+        $a; $a; $a; $a
+    end |> esc
+end
```
Comment (giordano) [inline base/sort.jl:1120]: Would `Cartesian.@nexprs 4` be enough instead of defining a specialised macro?

Comment (LSchwerdt) [reply base/sort.jl:1120]: Cartesian.@nexprs would work. I am not 100% happy with the requirement of nexprs to require an anonymous function. But I am not set on a specific implementation. Maybe there is an even better way to do it?

Here are some implementation possibilities to compare them directly:
```
while k <= hi - 4
    @repeat4x begin
        x = v[k]                  # lookup the element
        i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
        j = counts[i]             # lookup the target index
        t[j + offset] = x         # put the element where it belongs
        counts[i] = j + 1         # increment the target index for the next
        k += 1                    #  ↳ element in this bucket
    end
end
```

```
while k <= hi - 4
    Base.Cartesian.@nexprs 4 _ -> begin
        x = v[k]                  # lookup the element
        i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
        j = counts[i]             # lookup the target index
        t[j + offset] = x         # put the element where it belongs
        counts[i] = j + 1         # increment the target index for the next
        k += 1                    #  ↳ element in this bucket
    end
end
```

```
while k <= hi - 4
        x = v[k]                  # lookup the element
        i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
        j = counts[i]             # lookup the target index
        t[j + offset] = x         # put the element where it belongs
        counts[i] = j + 1         # increment the target index for the next
        k += 1                    #  ↳ element in this bucket
        x = v[k]                  
        i = (x >> shift)&mask + 1 
        j = counts[i]             
        t[j + offset] = x         
        counts[i] = j + 1         
        k += 1
        x = v[k]                  
        i = (x >> shift)&mask + 1 
        j = counts[i]             
        t[j + offset] = x         
        counts[i] = j + 1         
        k += 1 
        x = v[k]                  
        i = (x >> shift)&mask + 1 
        j = counts[i]             
        t[j + offset] = x         
        counts[i] = j + 1         
        k += 1               
end
```

Comment (LSchwerdt) [reply base/sort.jl]: Resolved by [cb97f9f](https://github.com/JuliaLang/julia/pull/48459/commits/cb97f9fd125fa45de56e775c5bff179765a6027f).

Comment (Seelengrab) [reply base/sort.jl]: I guess `@nexprs` could be special cased to allow a raw `begin` block for a 0-arg function as well. I have some plans for making the whole macro family of `Cartesian` a bit more useful/more user friendly, so I'll keep that in mind.

**Comment by LilithHafner** [comment]
Thanks!

> This makes the code somewhat less clean, but in my opinion the performance benefit is worth it.

50% speedup for 20 lines of code is definitely worth it!

I wish the Julia compiler could do this automatically and I'd like to add a `#TODO: remove this once the compiler is smart enough to do it automatically`, but in the interim (which might be a very long time) I support this PR for its immediate performance gains.

I'm having trouble reproducing your performance results. On my computer (2019 MacBook Air), pasting your script into a Julia REPL results in the following output:

```
base:     7.235 seconds
unrolled: 6.829 seconds
speedup factor: 1.06
```

<details>

<summary> I also tried directly measuring calls to `sort!`, but found no change with this script:</summary>

```julia
using BenchmarkTools, Random

function benchmark(k = 8)
    for n in 2:k-2
        println(n)
        v = rand(10^n)
        @btime sort!($v) setup=(rand!($v)) evals=1 gctrial=false gcsample=false samples=10^(k-n);
        v = rand(Int, 10^n)
        @btime sort!($v) setup=(rand!($v)) evals=1 gctrial=false gcsample=false samples=10^(k-n);
    end
end

benchmark()

@eval Base.Sort function radix_sort_pass!(t, lo, hi, offset, counts, v, shift, chunk_size)
    mask = UInt(1) << chunk_size - 1  # mask is defined in pass so that the compiler
    @inbounds begin                   #  ↳ knows it's shape
        # counts[2:mask+2] will store the number of elements that fall into each bucket.
        # if chunk_size = 8, counts[2] is bucket 0x00 and counts[257] is bucket 0xff.
        counts .= 0
        for k in lo:hi
            x = v[k]                  # lookup the element
            i = (x >> shift)&mask + 2 # compute its bucket's index for this pass
            counts[i] += 1            # increment that bucket's count
        end
        counts[1] = lo                # set target index for the first bucket
        cumsum!(counts, counts)       # set target indices for subsequent buckets
        # counts[1:mask+1] now stores indices where the first member of each bucket
        # belongs, not the number of elements in each bucket. We will put the first element
        # of bucket 0x00 in t[counts[1]], the next element of bucket 0x00 in t[counts[1]+1],
        # and the last element of bucket 0x00 in t[counts[2]-1].

		#loop unrolled 4x
        k = lo
        while k <= hi - 4
            Base.Cartesian.@nexprs 4 _ -> begin
                x = v[k]                  # lookup the element
                i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
                j = counts[i]             # lookup the target index
                t[j + offset] = x         # put the element where it belongs
                counts[i] = j + 1         # increment the target index for the next
                k += 1                    #  ↳ element in this bucket
            end
        end
        while k <= hi
            x = v[k]                  # lookup the element
            i = (x >> shift)&mask + 1 # compute its bucket's index for this pass
            j = counts[i]             # lookup the target index
            t[j + offset] = x         # put the element where it belongs
            counts[i] = j + 1         # increment the target index for the next
            k += 1                    #  ↳ element in this bucket
        end
    end
end

benchmark()
```

Output:

```
2
  1.779 μs (1 allocation: 896 bytes)
  1.532 μs (1 allocation: 896 bytes)
3
  15.251 μs (2 allocations: 10.12 KiB)
  15.976 μs (2 allocations: 10.12 KiB)
4
  180.667 μs (3 allocations: 86.36 KiB)
  199.008 μs (3 allocations: 86.36 KiB)
5
  1.866 ms (3 allocations: 789.48 KiB)
  2.106 ms (3 allocations: 789.48 KiB)
6
  23.867 ms (3 allocations: 7.64 MiB)
  27.281 ms (3 allocations: 7.64 MiB)
2
  1.794 μs (1 allocation: 896 bytes)
  1.567 μs (1 allocation: 896 bytes)
3
  15.819 μs (2 allocations: 10.12 KiB)
  16.955 μs (2 allocations: 10.12 KiB)
4
  180.848 μs (3 allocations: 86.36 KiB)
  192.534 μs (3 allocations: 86.36 KiB)
5
  1.778 ms (3 allocations: 789.48 KiB)
  2.069 ms (3 allocations: 789.48 KiB)
6
  23.991 ms (3 allocations: 7.64 MiB)
  27.122 ms (3 allocations: 7.64 MiB)
```

</details>

Time it takes to `sort!` a vector

Input | Before | After
---|---|---
rand(10^2) | 1.779 μs | 1.794 μs
rand(Int, 10^2) | 1.532 μs | 1.567 μs
rand(10^3) | 15.251 μs | 15.819 μs
rand(Int, 10^3) | 15.976 μs | 16.955 μs
rand(10^4) | 180.667 μs | 180.848 μs
rand(Int, 10^4) | 199.008 μs | 192.534 μs
rand(10^5) | 1.866 ms | 1.778 ms
rand(Int, 10^5) | 2.106 ms | 2.069 ms
rand(10^6) | 23.867 ms | 23.991 ms
rand(Int, 10^6) | 27.281 ms | 27.122 ms

**Comment by LilithHafner** [comment]
I doubt this will reveal much of note before https://github.com/JuliaCI/BaseBenchmarks.jl/pull/305 merges, but giving it a go anyway:

@nanosoldier `runbenchmarks("sort", vs=":master")`

**Comment by LSchwerdt** [comment]
`#TODO: remove this once the compiler is smart enough to do it automatically`
Yeah, I think you are right. It is not so much the amount of code, it is the feeling that it is somehow not necessary, because the compiler should be able to to this.

I did some more benchmarks using my test script on different hardware. It seems the speedup is not really reproducible. Maybe there is something wrong with my system? I think this should not be merged, unless someone else can reproduce the performance gain.

`n = 10^3, nRepeats = 10^8`
|  Hardware | time (base) | time (unrolled) | speedup factor |
| ------------- | ------------- | ------------- | ------------- |
| AMD 3900X | 156.316s | 164.073s | 0.95 |
| Intel 6700HQ | 210.218s | 243.852s | 0.86 |
| Intel 7820X | 174.933s | 193.171s | 0.91 |

` n = 10^7, nRepeats = 10^3`
|  Hardware | time (base) | time (unrolled) | speedup factor |
| ------------- | ------------- | ------------- | ------------- |
| AMD 3900X | 31.560s | 20.221s | 1.56 |
| Intel 6700HQ | 67.730s | 67.554s  | 1.00 |
| Intel  7820X | 38.566s | 39.933s | 0.97 |

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/48459#issuecomment-1409072236) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/8653d59_vs_a1c4d85/report.md).

**Comment by vtjnash** [comment]
> 50% speedup for 20 lines of code is definitely worth it!

The extra icache pressure can make this undesirable, even if it performs well in benchmarks, so it is not always better for the compiler to do this automatically where the user did not ask for it.

**Comment by LSchwerdt** [comment]
Thank you everyone. I reran the tests with different RAM configurations to close in on the reason for these performance discrepancies. And memory speeds made all the difference:

`n = 10^3, nRepeats = 2*10^7`
|  Hardware | time (base) | time (unrolled) | speedup factor |
| ------------- | ------------- | ------------- | ------------- |
| AMD 3900X, 2 DIMMs DDR4-**2333** | 31.051s | 31.999s | 0.97 |
| AMD 3900X, 4 DIMMs DDR4-**2333** | 31.173s | 32.346s | 0.86 |
| AMD 3900X, 4 DIMMs DDR4-**3600** | 33.012s | 32.536s | 1.01 |

` n = 10^7, nRepeats = 10^3`
|  Hardware | time (base) | time (unrolled) | speedup factor |
| ------------- | ------------- | ------------- | ------------- |
| AMD 3900X, 2 DIMMs DDR4-**2333** | 37.940s | 36.017s | 1.05 |
| AMD 3900X, 4 DIMMs DDR4-**2333** | 37.342s | 35.032s | 1.07 |
| AMD 3900X, 4 DIMMs DDR4-**3600** | 32.111s | 20.108s | 1.60 |

Is there any way to exploit this without hurting the performance on other hardware? I am sad to pass up this significant speedup opportunity, even though it is the right thing to do in light of the performance regressions for more common hardware configurations.

**Comment by LilithHafner** [comment]
Not a plausible near-term solution, but perhaps we could have ai tuned compiler heuristics that are trained separately on each detectable platform. (have to be AI tuned because there are more platforms than humans willing and able to tune compiler heuristics)

**Comment by LSchwerdt** [comment]
It would have been nice to get some performance numbers using more modern hardware (and more AMD CPUs), to see if this suboptimal performace without unrolled loop is more than a fluke. But without that, Im gonna close the PR. Writing an autotuning sort library is not my intention. ;)

If the standard library ever includes multithreaded (sort) functions this would be resolved automatically, because it is easier to saturate the memory bandwith using multiple threads. But maybe another algorithm with more cache-friendly memory accesses than LSD radix sort would be faster in that case anyways.

**Comment by Seelengrab** [comment]
> It would have been nice to get some performance numbers using more modern hardware (and more AMD CPUs), to see if this suboptimal performace without unrolled loop is more than a fluke.

I think I can help with that a bit:

```
[sukera@tower radixunroll]$ julia unroll.jl 
2000/20000
4000/20000
6000/20000
8000/20000
10000/20000
12000/20000
14000/20000
16000/20000
18000/20000
20000/20000

base:     1.892 seconds
unrolled: 2.152 seconds
speedup factor: 0.88

Julia Version 1.10.0-DEV.455
Commit 57f82bce39* (2023-01-30 10:26 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: 24 × AMD Ryzen 9 7900X 12-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, znver3)
  Threads: 24 on 24 virtual cores
Environment:
  JULIA_NUM_THREADS = 24
```

**Comment by LilithHafner** [comment]
@LSchwerdt, thank you for identifying this possible room for improvement, making a PR, and being receptive to feedback from others with different hardware that shed a different light on the performance. This sort of thing doesn't always pan out, but I encourage you to try again if you run across another opportunity for improvement.

**Pull request closed without merging**